rmk/src/keyboard.rs

```diff
@@ -51,21 +85,44 @@ impl<const ROW: usize, const COL: usize, const NUM_LAYER: usize, const NUM_ENCOD
     /// Main keyboard processing task, it receives input devices result, processes keys.
     /// The report is sent using `send_report`.
     async fn run(&mut self) {
+        let mut buffering = false;
         loop {
-            let key_event = KEY_EVENT_CHANNEL.receive().await;
-
-            // Process the key change
-            self.process_inner(key_event).await;
+            debug!("new loop, buffer: {:?}", self.holding_buffer);
+            //TODO add post time wait timer here
+            let hold_timeout_event = self.find_earliest_event();
+            let result: LoopState = match hold_timeout_event {
+                Some(event) => self.drain_unreleased_tap_hold_events(event, buffering).await,
+                _ => {
+                    // Process new key event
+                    let key_event = KEY_EVENT_CHANNEL.receive().await;
+                    // Process the key change
+                    self.process_inner(key_event).await
+                }
+            };
 
-            // After processing the key change, check if there are unprocessed events
-            // This will happen if there's recursion in key processing
-            loop {
-                if self.unprocessed_events.is_empty() {
-                    break;
+            buffering = false;
+            match result {
+                Queue => {
+                    // keep unprocessed key events
+                    // every key should be buffered into event list, check in every turn in future
+                    buffering = true;
+                    continue;
+                }
+                Stop => {
+                    return;
+                }
+                _ => {
```

Comment:
What's the difference between `LoopState::Flush` and `LoopState::Ok`? It seems that the two states can be merged